### PR TITLE
Improved 'stateless' helper-related classes

### PR DESCRIPTION
**Pull request by eduard-bagdasaryan**
No Squid functionality changes are expected beyond minor debugging
message changes.

### CODE REVIEW

**Comment by eduard-bagdasaryan** [comment]
These changes were [discussed](https://github.com/squid-cache/squid/pull/1389#discussion_r1237075585) and agreed on in a PR1389 discussion.

File: src/ssl/helper.cc [hunk 1]
```diff
@@ -72,7 +72,7 @@ operator <<(std::ostream &os, const Ssl::GeneratorRequest &gr)
 /// pending Ssl::Helper requests (to all certificate generator helpers combined)
 static Ssl::GeneratorRequests TheGeneratorRequests;
 
-helper::Pointer Ssl::Helper::ssl_crtd = nullptr;
+Helper::Client::Pointer Ssl::Helper::ssl_crtd = nullptr;
```
Comment (yadij) [inline src/ssl/helper.cc:75]: `Pointer` do not need explicit nullptr assignment.
```suggestion
Helper::Client::Pointer Ssl::Helper::ssl_crtd;
```

File: src/ssl/helper.cc [hunk 2]
```diff
@@ -166,7 +166,7 @@ Ssl::HandleGeneratorReply(void *data, const ::Helper::Reply &reply)
 }
 #endif //USE_SSL_CRTD
 
-helper::Pointer Ssl::CertValidationHelper::ssl_crt_validator = nullptr;
+Helper::Client::Pointer Ssl::CertValidationHelper::ssl_crt_validator = nullptr;
```
Comment (yadij) [inline src/ssl/helper.cc:169]: `Pointer` do not need explicit nullptr assignment.
```suggestion
Helper::Client::Pointer Ssl::CertValidationHelper::ssl_crt_validator;
```

File: src/helper.h [hunk 3]
```diff
@@ -221,7 +225,7 @@ class HelperServerBase: public CbdataParent
         bool shutdown;
     } flags;
 
-    typedef std::list<Helper::Xaction *> Requests;
+    typedef std::list<Xaction *> Requests;
```
Comment (yadij) [inline src/helper.h:228]:
```suggestion
    using Requests = std::list<Xaction *>;
```

File: src/helper.h [hunk 4]
```diff
@@ -32,6 +32,8 @@
 
 class Packable;
 class wordlist;
+class MemBuf;
+class CommTimeoutCbParams;
```
Comment (yadij) [inline src/helper.h:36]: alphabetical list please.
```suggestion
class CommTimeoutCbParams;
class MemBuf;
class Packable;
class wordlist;
```

File: src/helper.cc [hunk 5]
```diff
@@ -1439,14 +1434,14 @@ helperDispatch(helper_server * srv, Helper::Xaction * r)
     }
 
     r->request.Id = reqId;
-    helper_server::Requests::iterator it = srv->requests.insert(srv->requests.end(), r);
+    Helper::Session::Requests::iterator it = srv->requests.insert(srv->requests.end(), r);
```
Comment (yadij) [inline src/helper.cc:1437]: Can this use `auto` instead?
```suggestion
    const auto it = srv->requests.insert(srv->requests.end(), r);
```
or 
```suggestion
    auto it = srv->requests.insert(srv->requests.end(), r);
```

File: src/helper.cc [hunk 6]
```diff
@@ -1241,7 +1237,7 @@ helperStatefulHandleRead(const Comm::ConnectionPointer &conn, char *, size_t len
 
 /// Handles a request when all running helpers, if any, are busy.
 static void
-Enqueue(helper * hlp, Helper::Xaction * r)
+Enqueue(Helper::Client * const hlp, Helper::Xaction * const r)
```
Comment (yadij) [inline src/helper.cc:1240]: Please add a TODO about making `hlp` a (`Pointer`?) reference instead of raw-pointer

File: src/helper.cc [hunk 7]
```diff
@@ -931,10 +927,10 @@ helper_stateful_server::HelperServerClosed(helper_stateful_server *srv)
...
-helper_server::popRequest(int request_number)
+Helper::Session::popRequest(const int request_number)
 {
-    Helper::Xaction *r = nullptr;
-    helper_server::RequestIndex::iterator it;
+    Xaction *r = nullptr;
+    RequestIndex::iterator it;
     if (parent->childs.concurrency) {
         // If concurrency supported retrieve request from ID
         it = requestsIndex.find(request_number);
```
Comment (yadij) [inline src/helper.cc:936]: The iterator can be better scoped and probably use `auto` as well.
```suggestion
    if (parent->childs.concurrency) {
        // If concurrency supported retrieve request from ID
        const auto it = requestsIndex.find(request_number);
```

**Review by yadij — APPROVED** [review]
Just some polish, I do not insist on any of these.

File: src/ssl/helper.cc (hunk 1 above)
Comment (eduard-bagdasaryan) [reply src/ssl/helper.cc]: Done.

File: src/ssl/helper.cc (hunk 2 above)
Comment (eduard-bagdasaryan) [reply src/ssl/helper.cc]: Done.

**Review by eduard-bagdasaryan — COMMENTED** [review]
Done (ee124d5), please re-review.

**Review by rousskov — APPROVED** [review]
I pushed a few minor fixes and polishing touches (up to commit 3a080cc). No controversial changes AFAICT. There is a lot of (necessary) noise in this PR, so I could have easily missed something, but I think this PR is the right step forward that should unblock several improvements.

File: src/helper.h [hunk 8]
```diff
@@ -137,15 +137,17 @@ class helper: public RefCountable
     void submit(const char *buf, HLPCB * callback, void *data);
 };
 
-class statefulhelper : public helper
+} // namespace Helper
+
+// TODO: Rename to a *Client.
+class statefulhelper: public Helper::Client
```
[the remark this answers is not in the document]
Comment (yadij) [reply src/helper.h:143]: IMO this is an architectural decision, so for now just a TODO until the rest of the cleanup is further along.

**Review by yadij — APPROVED** [review]

**Pull request closed without merging**